src/Workspaces/Core/Portable/Rename/RenameSymbolAnnotation.cs

```diff
@@ -0,0 +1,107 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+#nullable enable 
+
+namespace Microsoft.CodeAnalysis.Rename
+{
+    /// <summary>
+    /// Provides a way to produce and consume annotations for rename
+    /// that contain the previous symbol serialized as a <see cref="SymbolKey" />.
+    /// This annotations is used by <see cref="Workspace.TryApplyChanges(Solution)" /> 
+    /// in some cases to notify the workspace host of refactorings.
+    /// </summary>
+    /// <remarks>
+    /// This annotation is applied to a symbol that has been renamed. 
+    /// When Workspace.TryApplyChanges happens in Visual Studio, we raise rename events for that symbol.
+    /// </remarks>
+    internal static class RenameSymbolAnnotation
+    {
+        public const string RenameSymbolKind = nameof(RenameSymbolAnnotation);
+
+        public static bool ShouldAnnotateSymbol(ISymbol symbol)
+            => symbol.DeclaringSyntaxReferences.Any();
+
+        public static SyntaxAnnotation? Create(ISymbol oldSymbol)
+        {
+            return ShouldAnnotateSymbol(oldSymbol)
+                ? new SyntaxAnnotation(RenameSymbolKind, SerializeData(oldSymbol))
+                : null;
+        }
+
+        public static TNode WithRenameSymbolAnnotation<TNode>(this TNode node, SemanticModel semanticModel)
+            where TNode : SyntaxNode
+            => node.WithAdditionalAnnotations(Create(semanticModel.GetDeclaredSymbol(node)));
+
+        public static async Task<ImmutableDictionary<ISymbol, ISymbol>> GatherChangedSymbolsInDocumentsAsync(
+            IEnumerable<DocumentId> changedDocumentIds,
+            Solution newSolution, Solution oldSolution,
+            CancellationToken cancellationToken)
+        {
+            var changedSymbols = ImmutableDictionary.CreateBuilder<ISymbol, ISymbol>();
+
+            foreach (var changedDocId in changedDocumentIds)
+            {
+                var newDocument = newSolution.GetDocument(changedDocId);
+
+                // Documents without syntax tree won't have the annotations attached
```

Comment:
You can be slightly stronger with SupportsSemanticModel. (We're wondering if for some remote cases we might have documents that support syntax because we can parse locally but have no semantics at all...)